src/main/java/io/confluent/connect/hdfs/DataWriter.java

```diff
@@ -0,0 +1,333 @@
+  private SinkTaskContext context;
+  private ExecutorService executorService;
+  private String hiveDatabase;
+  private HiveMetaStore hiveMetaStore;
+  private HiveUtil hive;
+  private Queue<Future> futures;
+  private boolean hiveIntegration;
+
+  @SuppressWarnings("unchecked")
+  public DataWriter(HdfsSinkConnectorConfig connectorConfig, SinkTaskContext context, AvroData avroData) {
+    try {
+      this.connectorConfig = connectorConfig;
+      this.avroData = avroData;
+      this.context = context;
+
+      conf = new Configuration();
+      url = connectorConfig.getString(HdfsSinkConnectorConfig.HDFS_URL_CONFIG);
+      topicsDir = connectorConfig.getString(HdfsSinkConnectorConfig.TOPICS_DIR_CONFIG);
+      String logsDir = connectorConfig.getString(HdfsSinkConnectorConfig.LOGS_DIR_CONFIG);
+
+      Class<? extends Storage> storageClass = (Class<? extends Storage>) Class
+              .forName(connectorConfig.getString(HdfsSinkConnectorConfig.STORAGE_CLASS_CONFIG));
+      storage = StorageFactory.createStorage(storageClass, conf, url);
+
+      createDir(topicsDir);
+      createDir(logsDir);
+
+      format = getFormat();
+      writerProvider = format.getRecordWriterProvider();
+      schemaFileReader = format.getSchemaFileReader(avroData);
+
+      partitioner = createPartitioner(connectorConfig);
+
+      assignment = context.assignment();
+      offsets = new HashMap<>();
+      lastAssignment = new HashSet<>();
+
+      hiveIntegration = connectorConfig.getBoolean(HdfsSinkConnectorConfig.HIVE_INTEGRATION_CONFIG);
+      if (hiveIntegration) {
+        hiveDatabase = connectorConfig.getString(HdfsSinkConnectorConfig.HIVE_DATABASE_CONFIG);
+        hiveMetaStore = new HiveMetaStore(conf, connectorConfig);
+        hive = format.getHiveUtil(connectorConfig, avroData, hiveMetaStore);
+        executorService = Executors.newSingleThreadExecutor();
+        futures = new LinkedList<>();
+      }
+
+      topicPartitionWriters = new HashMap<>();
+      for (TopicPartition tp: assignment) {
+        TopicPartitionWriter topicPartitionWriter = new TopicPartitionWriter(
+            tp, storage, writerProvider, partitioner, connectorConfig, context, avroData, hiveMetaStore, hive, schemaFileReader, executorService, futures);
+        topicPartitionWriters.put(tp, topicPartitionWriter);
+      }
+    } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+      throw new ConnectException("Reflection exception: ", e);
+    } catch (IOException e) {
+      throw new ConnectException(e);
+    }
+  }
+
+  public void write(Collection<SinkRecord> records) {
+    for (SinkRecord record: records) {
+      String topic = record.topic();
+      int partition = record.kafkaPartition();
+      TopicPartition tp = new TopicPartition(topic, partition);
+      topicPartitionWriters.get(tp).buffer(record);
+    }
+
+    if (hiveIntegration) {
+      Iterator<Future> iterator = futures.iterator();
```

Comment:
Would it make sense to do this in a callback tied to the future (e.g. see `FutureCallback` in Kafka Connect's utility library, that implements both a callback and a future)? It's not that big a deal to make it work as is, but it's basically polling on every `write` call rather than just being event-driven by removing the future immediately upon completion.

That said, maybe the current approach is just simpler since it avoids any additional locking? I assume the future is completed in the `executorService` thread.